flaky test: TestAWSProvider_GetDefaultRegion leaks ambient AWS_REGION via IsConfigured()

Discovered while running the providers/aws test suite during the pre-flight for #85 / PR #94.

## Symptom

`go test ./providers/aws/...` fails on a developer machine that has `~/.aws/config` with a default region (or `AWS_REGION` set in the env):

```
--- FAIL: TestAWSProvider_GetDefaultRegion/Config_region_set_(no_provider_region)
    Error: Not equal:
    expected: "ap-southeast-1"
    actual  : "ap-south-1"        # leaked from ~/.aws/config [default] profile
```

CI may pass because the runner has no `[default]` AWS profile / no `AWS_REGION` env, but the test is non-deterministic across environments.

## Root cause

`providers/aws/provider.go::AWSProvider.GetDefaultRegion` calls `p.IsConfigured()`, which calls `p.cfgOnce.Do(p.loadConfig)`. `loadConfig` invokes the SDK default credentials/config chain (env → `~/.aws/config` → IMDS). That chain **overwrites `p.cfg.Region`** that the test had set explicitly, so the test assertion fails when the ambient region differs from the test fixture.

## Steps to reproduce

1. Set `[default]\nregion = ap-south-1` in `~/.aws/config` (or `export AWS_REGION=ap-south-1`).
2. Run `go test -run TestAWSProvider_GetDefaultRegion ./providers/aws/`.

## Expected

`GetDefaultRegion()` should return what the test set on `p.cfg.Region` without consulting the ambient environment.

## Proposed fix

Either:

- (a) Decouple `GetDefaultRegion` from `IsConfigured` — return `p.cfg.Region` if it is non-empty without triggering `loadConfig`. Only fall back to the SDK chain when `cfg` was never set.
- (b) In the test, isolate the ambient env: `t.Setenv("AWS_REGION", "")`, `t.Setenv("AWS_DEFAULT_REGION", "")`, `t.Setenv("AWS_CONFIG_FILE", "/dev/null")`, `t.Setenv("AWS_SHARED_CREDENTIALS_FILE", "/dev/null")` before the test cases that pre-set `cfg.Region`. Cleaner is (a) — the production code shouldn’t silently override caller-supplied region.

## Severity

**Low** — the production code path works correctly (the SDK fallback is desirable when no region was supplied), but the test is flaky across developer machines and has no env isolation. Same root cause likely affects the `TestMemorystoreClient_GetExistingCommitments_WithMockService` failures observed in `providers/gcp/services/memorystore/client_test.go` — worth a separate audit pass over `providers/**/_test.go` for similar ambient-env leaks.

## References

Discovered during pre-flight for PR #94 / issue #85.

--- a/providers/aws/provider.go
+++ b/providers/aws/provider.go
@@ -378,10 +378,31 @@ func (p *AWSProvider) GetRegions(ctx context.Context) ([]common.Region, error) {
 	return regions, nil
 }
 
-// GetDefaultRegion returns the default AWS region
+// GetDefaultRegion returns the default AWS region.
+//
+// Priority order:
+//
+//  1. p.region if explicitly set on the provider.
+//  2. p.cfg.Region if the caller pre-populated cfg (e.g. tests with a
+//     synthetic config) — checked BEFORE IsConfigured to avoid the SDK's
+//     credentials/config chain overwriting a caller-supplied region with
+//     the ambient AWS_REGION / ~/.aws/config default.
+//  3. The SDK-loaded region (via IsConfigured → loadConfig).
+//  4. "us-east-1" as a last-resort default.
+//
+// Step 2 is what fixes #96: previously this function called
+// IsConfigured() unconditionally, which triggers cfgOnce.Do(loadConfig)
+// and lets the SDK's credentials/config chain populate p.cfg.Region from
+// the developer's ambient environment, masking a region the test had set
+// explicitly. Reading the already-set p.cfg.Region first keeps test
+// fixtures honest while still falling through to the SDK chain when no
+// caller has touched cfg.
 func (p *AWSProvider) GetDefaultRegion() string {
 	if p.region != "" {
 		return p.region
+	}
+	if p.cfg.Region != "" {
+		return p.cfg.Region
 	}
 	if p.IsConfigured() && p.cfg.Region != "" {
 		return p.cfg.Region

--- a/providers/aws/provider_test.go
+++ b/providers/aws/provider_test.go
@@ -158,6 +158,13 @@ func TestAWSProvider_DisplayName(t *testing.T) {
 }
 
 func TestAWSProvider_GetDefaultRegion(t *testing.T) {
+	// Isolate from ambient AWS env on the developer machine — without this,
+	// the SDK's credentials/config chain (loadConfig via IsConfigured) would
+	// overwrite p.cfg.Region with whatever the developer's ~/.aws/config or
+	// AWS_REGION env var says, masking the values these test cases set
+	// explicitly. See #96.
+	clearAWSAmbientEnv(t)
+
 	tests := []struct {
 		name           string
 		provider       *AWSProvider
@@ -195,6 +202,39 @@ func TestAWSProvider_GetDefaultRegion(t *testing.T) {
 			assert.Equal(t, tt.expectedRegion, tt.provider.GetDefaultRegion())
 		})
 	}
+}
+
+// TestAWSProvider_GetDefaultRegion_NoLeakViaIsConfigured is the regression
+// guard for #96: GetDefaultRegion must NOT trigger the SDK's
+// credentials/config chain when the caller has already populated
+// p.cfg.Region. This prevents ambient AWS_REGION / ~/.aws/config from
+// leaking in and silently overwriting test-supplied regions.
+//
+// We force a non-default ambient region and confirm GetDefaultRegion
+// returns the test-supplied region, not the ambient one.
+func TestAWSProvider_GetDefaultRegion_NoLeakViaIsConfigured(t *testing.T) {
+	t.Setenv("AWS_REGION", "ap-south-1")
+	t.Setenv("AWS_DEFAULT_REGION", "ap-south-1")
+	// Don't clear AWS_CONFIG_FILE here — the test still works if one exists,
+	// because our fix consults p.cfg.Region BEFORE IsConfigured.
+
+	p := &AWSProvider{
+		cfg: aws.Config{Region: "ap-southeast-1"},
+	}
+	got := p.GetDefaultRegion()
+	assert.Equal(t, "ap-southeast-1", got, "test-supplied cfg.Region must not be overwritten by ambient AWS_REGION")
+}
+
+// clearAWSAmbientEnv zeroes the ambient AWS env vars + redirects the
+// SDK's config/credentials file lookups to /dev/null, isolating tests
+// that exercise loadConfig from developer machine state.
+func clearAWSAmbientEnv(t *testing.T) {
+	t.Helper()
+	t.Setenv("AWS_REGION", "")
+	t.Setenv("AWS_DEFAULT_REGION", "")
+	t.Setenv("AWS_PROFILE", "")
+	t.Setenv("AWS_CONFIG_FILE", "/dev/null")
+	t.Setenv("AWS_SHARED_CREDENTIALS_FILE", "/dev/null")
 }
 
 func TestAWSProvider_GetSupportedServices(t *testing.T) {

